File: bank-service/src/main/java/rs/edu/raf/mapper/CreditRequestMapper.java
package rs.edu.raf.mapper;


import lombok.AllArgsConstructor;
import org.springframework.stereotype.Component;
import rs.edu.raf.dto.CreditRequestCreateDto;
import rs.edu.raf.dto.CreditRequestDTO;
import rs.edu.raf.model.credit.CreditRequest;
import rs.edu.raf.model.credit.CreditRequestStatus;
import rs.edu.raf.model.credit.CreditType;
import rs.edu.raf.repository.CurrencyRepository;
import rs.edu.raf.repository.accounts.BankAccountRepository;

@Component
@AllArgsConstructor
public class CreditRequestMapper {
    private BankAccountRepository bankAccountRepository;
    private CurrencyRepository currencyRepository;
    private CurrencyMapper currencyMapper;
    private BankAccountMapper bankAccountMapper;
    private CreditTypeMapper creditTypeMapper;

    public CreditRequest creditRequestCreateDtoToCreditRequest(CreditRequestCreateDto creditRequestCreateDto, CreditType creditType){
        CreditRequest creditRequest = new CreditRequest();

        creditRequest.setCreditType(creditType);
        creditRequest.setLoanAmount(creditRequestCreateDto.amount());
        creditRequest.setCurrency(currencyRepository.findById(creditRequestCreateDto.currencyId()).orElseThrow());
        creditRequest.setSalary(creditRequestCreateDto.salary());
        creditRequest.setLoanPurpose(creditRequestCreateDto.loanPurpose());
        creditRequest.setPhoneNumber(creditRequestCreateDto.phoneNumber());
        creditRequest.setBankAccount(bankAccountRepository.findByAccountNumber(creditRequestCreateDto.bankAccountNumber()).orElseThrow());
        creditRequest.setLoanTerm(creditRequestCreateDto.loanTerm());
        creditRequest.setPermanentEmployee(creditRequestCreateDto.permanentEmployee());
        creditRequest.setCurrentEmploymentPeriod(creditRequestCreateDto.currentEmploymentPeriod());
        creditRequest.setCreditRequestStatus(CreditRequestStatus.PENDING);

        return creditRequest;
    }

    public CreditRequestDTO creditRequestToCreditRequestDto(CreditRequest creditRequest){
        return new CreditRequestDTO(creditRequest.getId(), creditTypeMapper.creditTypeToCreditTypeDTO(creditRequest.getCreditType()),
                creditRequest.getLoanAmount(), currencyMapper.currencyToCurrencyDTO(creditRequest.getCurrency()),
                creditRequest.getLoanPurpose(), creditRequest.getSalary(), creditRequest.getPhoneNumber(),
                creditRequest.isPermanentEmployee(), creditRequest.getCurrentEmploymentPeriod(),
                creditRequest.getLoanTerm(), bankAccountMapper.bankAccountToBankAccountDTO(creditRequest.getBankAccount()),
                creditRequest.getCreditRequestStatus().toString());
    }
}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/mapper/CreditRequestMapper.kt
package org.example.bankservicekotlin.mapper

import org.example.bankservicekotlin.dto.CreditRequestCreateDTO
import org.example.bankservicekotlin.dto.CreditRequestDTO
import org.example.bankservicekotlin.model.credit.CreditRequest
import org.example.bankservicekotlin.model.credit.CreditRequestStatus
import org.example.bankservicekotlin.model.credit.CreditType
import org.example.bankservicekotlin.repository.CurrencyRepository
import org.example.bankservicekotlin.repository.accounts.BankAccountRepository
import org.springframework.stereotype.Component

@Component
class CreditRequestMapper(
    private val bankAccountRepository: BankAccountRepository,
    private val currencyRepository: CurrencyRepository,
    private val currencyMapper: CurrencyMapper,
    private val bankAccountMapper: BankAccountMapper,
    private val creditTypeMapper: CreditTypeMapper
) {

    fun creditRequestCreateDtoToCreditRequest(
        creditRequestCreateDto: CreditRequestCreateDTO,
        creditType: CreditType
    ): CreditRequest {
        val bankAccount =
            bankAccountRepository.findByAccountNumber(creditRequestCreateDto.bankAccountNumber).orElseThrow()
        return CreditRequest(
            null,
            null,
            creditType,
            creditRequestCreateDto.amount,
            creditRequestCreateDto.loanPurpose,
            creditRequestCreateDto.salary,
            creditRequestCreateDto.phoneNumber,
            creditRequestCreateDto.permanentEmployee,
            creditRequestCreateDto.currentEmploymentPeriod,
            creditRequestCreateDto.loanTerm,
            bankAccount,
            CreditRequestStatus.PENDING,
            null
        )
    }

    fun creditRequestToCreditRequestDto(creditRequest: CreditRequest): CreditRequestDTO {
        return CreditRequestDTO(
            creditRequest.id, creditTypeMapper.creditTypeToCreditTypeDTO(creditRequest.creditType),
            creditRequest.loanAmount, creditRequest.loanPurpose, creditRequest.salary,
            creditRequest.phoneNumber, creditRequest.permanentEmployee,
            creditRequest.currentEmploymentPeriod, creditRequest.loanTerm,
            bankAccountMapper.bankAccountToBankAccountDTO(creditRequest.bankAccount),
            creditRequest.creditRequestStatus.toString()
        )
    }
}